File: Matrix&Glist/Glist/Glist.py
class Glist:

    def __init__(self, tag):
        self.tag = tag          #1: sublist 0 : atom
        self.atom = None
        self.right = None
        self.down = None
        pass

    pass

class Solution:
    def __substr__(self, Str, l, r):
        res = ''
        for i in range(l, r):
            res = res + Str[i]
        return res
        pass
    def __split__(self, Str):
        cnt, pos = 0, 0
        while pos < len(Str) and (Str[pos] != ',' or cnt != 0):
            if Str[pos] == '(':
                cnt += 1
            if Str[pos] == ')':
                cnt -= 1
            pos += 1
        l = ''
        r = ''
        for i in range(len(Str)):
            if i < pos:
                l = l + Str[i]
            if i > pos:
                r = r + Str[i]
        #r = ''
        return l, r

    @classmethod
    def classMethod(cls):
        print 'Class'

    @staticmethod
    def method():
        print 'Static'


    def solve(self, Str):
        l = Glist(1)
        if len(Str) == 1:
            l.tag, l.atom = 0, Str
            return l
        content = self.__substr__(Str, 1, len(Str) - 1)
        if content == '':
            return None
        else:
            p = l
            left, right = self.__split__(content)
            p.down = self.solve(left)
            q = Glist(1)
            if right != '':
                p.right = q
                p = q
            while right != '':
                left, right = self.__split__(right)
                p.down = self.solve(left)
                if right != '':
                    q = Glist(1)
                    p.right = q
                    p = q
            p.right = None
            pass
        pass
        return l


'''
@property
1. private access
2. easy access
3. can't be modified
'''


def Print(self):
    if self is None or self.tag == 0:
        if self is None:
            print '()',
        else:
            print '%s' % self.atom,
        return
    print '(',
    cur = self
    while cur is not None:
        Print(cur.down)
        cur = cur.right
        if cur is not None:
            print ',',
        pass
    print ')',
    pass


def depth(l):
    if l is None or l.tag == 0:
        return 0
    mx = 0
    while l is not None:
        mx = max(mx, depth(l.down))
        l = l.right
    return mx + 1


if __name__ == '__main__':
    l = Solution().solve('(a,(),(b,c))')
    Print(l)
    print
    print(depth(l))
    pass

File: Matrix&Glist/Glist/Glist.cpp
#include <bits/stdc++.h>
using namespace std;
typedef char AtomType;
struct Glist{
    bool tag;
    union{
        AtomType atom;
        struct Node{
            Glist * hp;
            Glist * tp;
        }ptr;
    }u;
    Glist(bool tag = true){
        this->tag = tag;
    }
};

namespace Solution{
    void split(string &l, string &s){
        int k = 0;
        int pos = 0;
        do{
            if(s[pos] == '(') ++k;
            if(s[pos] == ')') --k;
            ++pos;
        }while(pos < s.length() && (s[pos] != ',' || k != 0));
        l = s.substr(0, pos);
        if(pos < s.length() - 1) s = s.substr(pos + 1);
        else s = "";
    }
    Glist * solve(string s){
        if(s == "()") return NULL;
        Glist * l = new Glist();
        if(s.length() == 1){
            l->tag = false;
            l->u.atom = s[0];
            return l;
        }
        s = s.substr(1, s.length() - 2);
        string left;
        Glist * p = l;
        do{
            //cout << "F: " << s << endl;
            split(left, s);
            //cout << left << " " << s << endl;
            p->u.ptr.hp = solve(left);
            if(s != ""){
                Glist * q = new Glist();
                p->u.ptr.tp = q;
                p = q;
            }
        }while(s != "");
        p->u.ptr.tp = NULL;
        //s.su
        return l;
    }

    int length(Glist * l){
        int len = 0;
        while(l){
            ++len;
            l = l->u.ptr.tp;
        }
        return len;
    }

    int depth(Glist * l){
        if(l == NULL || l->tag == false) return 0;
        int mx = 0;
        for(Glist * cur = l; cur; cur = cur->u.ptr.tp)
            mx = max(mx, depth(cur->u.ptr.hp));
        return mx + 1;
    }
}

void Print(Glist * x){
    if(x == NULL || x->tag == false){
        if(x == NULL) printf("()");
        else printf("%c", x->u.atom);
        return;
    }
    printf("(");
    while(x){
        Print(x->u.ptr.hp);
        x = x->u.ptr.tp;
        if(x) printf(",");
    }
    printf(")");
}

int main(){
    Glist * l = Solution::solve("((),a,(b,c),(b,(d,c,b)))");
    Print(l);
    printf("\n%d\n", Solution::length(l));
    printf("%d\n", Solution::depth(l));
    return 0;
}
